src/Microsoft.ML.Dnn/DnnTransform.cs

```diff
@@ -0,0 +1,1737 @@
+
+                            var editor = VBufferEditor.Create(ref dst, (int)tensorSize);
+
+                            DnnUtils.FetchData<T>(tensor.Data<T>(), editor.Values);
+                            dst = editor.Commit();
+                        };
+                        return valuegetter;
+                    }
+                }
+            }
+
+            private void UpdateCacheIfNeeded(long position, ITensorValueGetter[] srcTensorGetters, string[] activeOutputColNames, OutputCache outputCache)
+            {
+                if (outputCache.Position != position)
+                {
+                    bool addToTheBag = false;
+                    if (!_runners.TryTake(out Runner runner))
+                    {
+                        runner = _parent.Runner.CloneRunner();
+                        addToTheBag = true;
+                    }
+                    for (int i = 0; i < _inputColIndices.Length; i++)
+                    {
+                        var inputName = _parent.Inputs[i];
+                        runner.AddInput(i, srcTensorGetters[i].GetTensor());
+                    }
+
+                    var tensors = runner.Run();
+                    Contracts.Assert(tensors.Length > 0);
+
+                    for (int j = 0; j < activeOutputColNames.Length; j++)
+                        outputCache.Outputs[activeOutputColNames[j]] = tensors[j];
+
+                    outputCache.Position = position;
+
+                    if (addToTheBag)
+                        _runners.Add(runner);
+                }
+            }
+
+            private protected override Func<int, bool> GetDependenciesCore(Func<int, bool> activeOutput)
+            {
+                return col => Enumerable.Range(0, _parent.Outputs.Length).Any(i => activeOutput(i)) && _inputColIndices.Any(i => i == col);
+            }
+
+            protected override DataViewSchema.DetachedColumn[] GetOutputColumnsCore()
+            {
+                var info = new DataViewSchema.DetachedColumn[_parent.Outputs.Length];
+                for (int i = 0; i < _parent.Outputs.Length; i++)
+                    info[i] = new DataViewSchema.DetachedColumn(_parent.Outputs[i], _parent.OutputTypes[i], null);
+                return info;
+            }
+        }
+
+        private interface ITensorValueGetter
+        {
+            Tensor GetTensor();
+
+            void BufferTrainingData();
+
+            Tensor GetBufferedBatchTensor();
+        }
+
+        private class TensorValueGetter<T> : ITensorValueGetter
+        {
+            private readonly ValueGetter<T> _srcgetter;
+            private readonly T[] _bufferedData;
+            private readonly Int64[] _bufferedDataLong;
+            private readonly TensorShape _tfShape;
+            private int _position;
+            private readonly bool _keyType;
+
+            public TensorValueGetter(
```

Comment:
Same comment I did when using the `Options` class:

This class name (`Options`) feels too generic. 

- If it is specific and exclusive about **TensorFlow** it should probably be named as 'TensorFlowOptions'. 

- If it'll be a common type shared among TF and Torch, then maybe call it `DnnModelOptions`.

If it is a shared/generic class, I see potential issues here. This type looks generic since it is part of the DnnEstimator constructor, however, the internal options look very specific to TensorFlow. I don't think you wil have properties such as `TensorFlowLabel` and not sure that all the rest of properties of Options would be applicable to Torch?

Reply:
replied earlier on this topic.

---
In reply to: [310295321](https://github.com/dotnet/machinelearning/pull/4057#discussion_r310295321) [](ancestors = 310295321)